routes.py

```
@@ -17,6 +17,57 @@
+        if _ENTITY_DECL_TEXT_RE.search(text):
+            raise ValueError(
+                'XML file declares a custom entity, which is not permitted '
+                '(entity-expansion / "billion laughs" protection).'
+            )
+
+
+def _safe_parse_xml_file(path):
+    """`ET.parse(path)` with the entity-expansion guard applied first."""
+    xml_bytes = Path(path).read_bytes()
+    _reject_entity_declarations(xml_bytes)
+    return ET.parse(path)
```

Ruff (0.16.1)

[error] Using `xml` to parse untrusted data is known to be vulnerable to XML attacks; use `defusedxml` equivalents

(S314)